### PR TITLE
Add remote GPU processing time collection and forwarding

### PR DESCRIPTION
**Pull request by hansent**
## Summary

- Adds a thread-safe `RemoteProcessingTimeCollector` in `inference_sdk/config.py` that accumulates `(model_id, processing_time)` tuples from GPU instance responses via a `ContextVar`
- Collects `X-Processing-Time` headers from all remote HTTP calls in the SDK — both the `execute_requests_packages` funnel and the 3 direct `requests.post()` calls (CLIP text, CLIP compare, perception encoder)
- Propagates the collector through both ThreadPoolExecutor boundaries: workflow step executor and `run_in_parallel` utility
- Returns two new response headers from `GCPServerlessMiddleware`: `X-Remote-Processing-Time` (total sum) and `X-Remote-Processing-Times` (JSON array with per-model breakdown)
- Gated behind `WORKFLOWS_REMOTE_EXECUTION_TIME_FORWARDING` env var (default True)
- Includes 30 unit tests covering the collector, SDK collection points, ContextVar propagation, and middleware output

## Test plan

- [x] `RemoteProcessingTimeCollector` unit tests (add, get_total, get_entries, has_data, thread safety, default model_id, copy safety)
- [x] `_extract_model_id_from_request_data` tests (payload extraction, URL path fallback)
- [x] `_collect_remote_processing_times` tests (collection, skipping missing/malformed headers, mismatched counts)
- [x] `_collect_processing_time_from_response` tests (with/without collector, model_id, edge cases)
- [x] `run_in_parallel` ContextVar propagation tests (collector + execution_id into threads)
- [x] `GCPServerlessMiddleware` integration tests (headers present/absent, forwarding enabled/disabled)
- [ ] E2E: deploy with `WORKFLOWS_REMOTE_EXECUTION_TIME_FORWARDING=True` + `WORKFLOWS_STEP_EXECUTION_MODE=remote`, run a workflow, verify both new headers

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: inference_sdk/config.py [hunk 1]
```diff
@@ -1,10 +1,44 @@
...
+        with self._lock:
+            self._entries.append((model_id, processing_time))
+
+    def get_total(self) -> float:
+        with self._lock:
+            return sum(t for _, t in self._entries)
+
+    def get_entries(self) -> list:
+        with self._lock:
+            return list(self._entries)
```
Comment (grzegorz-roboflow) [inline inference_sdk/config.py]: looks like subsequent calls to `get_entries` with interleaved `add` might result in accumulated entries, should we clear self._entries here?

File: inference_sdk/config.py [hunk 2]
```diff
@@ -1,10 +1,44 @@
...
+        self._entries: list = []  # list of (model_id, time) tuples
+        self._lock = threading.Lock()
+
+    def add(self, processing_time: float, model_id: str = "unknown") -> None:
+        with self._lock:
+            self._entries.append((model_id, processing_time))
+
+    def get_total(self) -> float:
+        with self._lock:
+            return sum(t for _, t in self._entries)
```
Comment (grzegorz-roboflow) [inline inference_sdk/config.py]: similar observation as with get_entries - subsequent call interleaved with `add` will result in accumulated entries

File: inference_sdk/config.py [hunk 3]
```diff
@@ -1,10 +1,44 @@
...
+    def __init__(self):
+        self._entries: list = []  # list of (model_id, time) tuples
+        self._lock = threading.Lock()
+
+    def add(self, processing_time: float, model_id: str = "unknown") -> None:
+        with self._lock:
+            self._entries.append((model_id, processing_time))
+
+    def get_total(self) -> float:
+        with self._lock:
```
Comment (grzegorz-roboflow) [inline inference_sdk/config.py]: I wonder how will this lock behave in async

File: inference_sdk/http/client.py [hunk 4]
```diff
@@ -105,6 +110,21 @@
...
+) -> None:
+    collector = remote_processing_times.get()
+    if collector is None:
+        return
+    pt = response.headers.get(PROCESSING_TIME_HEADER)
+    if pt is not None:
+        try:
+            collector.add(float(pt), model_id=model_id)
+        except (ValueError, TypeError):
+            pass
```
Comment (grzegorz-roboflow) [inline inference_sdk/http/client.py]: I'd log with at least warning here

File: inference_sdk/http/utils/executors.py [hunk 5]
```diff
@@ -57,17 +62,55 @@ def execute_requests_packages(
...
+        if pt is not None:
+            model_id = (
+                _extract_model_id_from_request_data(requests_data[i])
+                if i < len(requests_data)
+                else "unknown"
+            )
+            try:
+                collector.add(float(pt), model_id=model_id)
+            except (ValueError, TypeError):
+                pass
```
Comment (grzegorz-roboflow) [inline inference_sdk/http/utils/executors.py]: we might want to log warning here (or set flag and log outside of `for` once)

File: inference/core/interfaces/http/http_api.py [hunk 6]
```diff
@@ -292,17 +301,39 @@ async def dispatch(self, request, call_next):
...
+        if collector is not None and collector.has_data():
+            response.headers[REMOTE_PROCESSING_TIME_HEADER] = str(
+                collector.get_total()
+            )
+            response.headers[REMOTE_PROCESSING_TIMES_HEADER] = json.dumps(
+                [
+                    {"model_id": model_id, "time": t}
+                    for model_id, t in collector.get_entries()
+                ]
+            )
```
Comment (grzegorz-roboflow) [inline inference/core/interfaces/http/http_api.py]: depending on proxy/server header size limits (4-8KB) and for big enough workflow we reach that limit if we store full json

Comment (grzegorz-roboflow) [reply inference/core/interfaces/http/http_api.py]: also, if we want to capture total and individual entries we might get inconsistent results if `add` is called between `get_entries` and `get_total` - we could `get_entries` and calculate total from that

File: inference/core/env.py [hunk 7]
```diff
@@ -289,6 +289,10 @@
 # Whether is's GCP serverless service
 GCP_SERVERLESS = str2bool(os.getenv("GCP_SERVERLESS", "False"))
 
+WORKFLOWS_REMOTE_EXECUTION_TIME_FORWARDING = str2bool(
+    os.getenv("WORKFLOWS_REMOTE_EXECUTION_TIME_FORWARDING", "True")
+)
```
Comment (grzegorz-roboflow) [inline inference/core/env.py:294]: I know we keep "X-Processing-Time" in constants, but I realized in sdk we keep it in env - maybe we can move `PROCESSING_TIME_HEADER` from constants here

Comment (hansent) [reply inference/core/env.py:294]: not sure I understand what you mean.  This isn't the header name but a flag to turn whether it does the remote exec collection / returns the new header.  we don't have flag for the other one, its always on. You mean to make the the declaration of REMOTE_PROCESSING_TIME_HEADER = `X-Remote-Processing-Time` into constants (its currently in http_api module scope)

File: inference/core/workflows/core_steps/common/utils.py [hunk 8]
```diff
@@ -463,9 +463,29 @@ def post_process_ocr_result(
 
 
 def run_in_parallel(tasks: List[Callable[[], T]], max_workers: int = 1) -> List[T]:
```
Comment (dkosowski87) [inline inference/core/workflows/core_steps/common/utils.py:465]: Not to be picky but the function uses `ThreadPoolExecutor` but it is named `run_in_parallel`. Would change it to `run_concurrent` or `run_with_multithreading`, as it's not true parallelism. One could expect looking at this function name that there is a Process pool underneath.

File: inference/core/workflows/core_steps/common/utils.py [hunk 9]
```diff
@@ -463,9 +463,29 @@ def post_process_ocr_result(
...
-    with ThreadPoolExecutor(max_workers=max_workers) as executor:
-        return list(executor.map(_run, tasks))
+    try:
+        from inference_sdk.config import execution_id, remote_processing_times
```
Comment (dkosowski87) [inline inference/core/workflows/core_steps/common/utils.py:479]: Is this inherently connected to each core step? If we have steps that don't need the processing time stats returned then perhaps it's better to wrap the incoming function with `run_with_context` externally or better yet provide a `with` context like (see example below). This would reduce coupling and leave the standard multithreading util function for usage elsewhere.

```python

def run_with_multithreading(
    tasks: List[Callable[[], T]],
    max_workers: int = 1,
) -> List[T]:
    """
    Pure multithreading executor.
    No knowledge of inference_sdk or context propagation.
    """
    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        return list(executor.map(lambda f: f(), tasks))


@contextmanager
def propagate_inference_context() -> Iterator[Callable[[Callable[[], T]], Callable[[], T]]]:
    """
    Captures execution_id and remote_processing_times (if available)
    and returns a wrapper factory that propagates them into worker threads.
    """
    try:
        from inference_sdk.config import execution_id, remote_processing_times

        exec_id = execution_id.get() if execution_id is not None else None
        collector = (
            remote_processing_times.get()
            if remote_processing_times is not None
            else None
        )
    except ImportError:
        exec_id = None
        collector = None

    def wrap(fun: Callable[[], T]) -> Callable[[], T]:
        def wrapped() -> T:
            if exec_id is not None:
                from inference_sdk.config import execution_id
                execution_id.set(exec_id)

            if collector is not None:
                from inference_sdk.config import remote_processing_times
                remote_processing_times.set(collector)

            return fun()

        return wrapped

    yield wrap

    with propagate_inference_context() as wrap:
        wrapped_tasks = [wrap(task) for task in tasks]
        return run_with_multithreading(wrapped_tasks, max_workers)
```

File: inference_sdk/http/utils/executors.py [hunk 10]
```diff
@@ -57,17 +62,57 @@ def execute_requests_packages(
...
+        model_id = request_data.payload.get("model_id")
+        if model_id:
+            return str(model_id)
+    try:
+        from urllib.parse import urlparse
+
+        path = urlparse(request_data.url).path
+        return path.strip("/")
+    except Exception:
+        return "unknown"
```
Comment (dkosowski87) [inline inference_sdk/http/utils/executors.py]: Let's move it to a constant with a proper naming `EXCEPTION_RAISED_MODEL_ID_PLACEHOLDER`, it's also used in `_collect_remote_processing_times`

File: inference_sdk/http/utils/executors.py [hunk 11]
```diff
@@ -57,17 +62,57 @@ def execute_requests_packages(
...
+    collector = remote_processing_times.get()
+    if collector is None:
+        return
+    for i, response in enumerate(responses):
+        pt = response.headers.get(PROCESSING_TIME_HEADER)
+        if pt is not None:
+            model_id = (
+                _extract_model_id_from_request_data(requests_data[i])
+                if i < len(requests_data)
+                else "unknown"
```
Comment (dkosowski87) [inline inference_sdk/http/utils/executors.py]: Do you really want to have a mismatch between the `requests_data` length and the collectors stats model_id length? Perhaps we just should do a `break` in a situation when the number of responses is larger than the number or requests.

File: inference/core/workflows/core_steps/common/utils.py (hunk 8 above)
Comment (hansent) [reply inference/core/workflows/core_steps/common/utils.py:465]: Yeah. I will not change apart of this PR, but will want to look into this anyway soon. Benchmarking on async deployment where it does the remote requests should be going through this, but I am seeing response times get longer with what looks at same rate as number of model blocks I have in one workflow (need to confirm configuration and run more tests today still though)
